// @prisma/engine-core/dist/esm/data-proxy/DataProxyEngine.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
import { Engine } from "../common/Engine";
import { request } from "./utils/request";
import EventEmitter from "events";
import { createSchemaHash } from "./utils/createSchemaHash";
import { backOff } from "./utils/backOff";
import { getClientVersion } from "./utils/getClientVersion";
const randomDebugId = Math.ceil(Math.random() * 1e3);
const MAX_RETRIES = 5;
class DataProxyEngine extends Engine {
  constructor(config) {
    super();
    this.config = config;
    this.env = this.config.env ?? {};
    this.inlineSchema = config.inlineSchema ?? "";
    this.inlineDatasources = config.inlineDatasources ?? {};
    this.logEmitter = new EventEmitter();
    this.logEmitter.on("error", () => {
    });
    this.initPromise = this.init();
  }
  async init() {
    const [host, apiKey] = this.extractHostAndApiKey();
    this.schemaHash = await createSchemaHash(this.inlineSchema);
    this.clientVersion = getClientVersion(this.config);
    this.headers = { Authorization: `Bearer ${apiKey}` };
    this.host = host;
  }
  version() {
    return "unknown";
  }
  async start() {
  }
  async stop() {
  }
  on(event, listener) {
    if (event === "beforeExit") {
      throw new Error("beforeExit event is not supported yet");
    } else {
      this.logEmitter.on(event, listener);
    }
  }
  async url(s) {
    await this.initPromise;
    return `https://${this.host}/${this.clientVersion}/${this.schemaHash}/${s}?id=${randomDebugId}`;
  }
  async getConfig() {
    await this.initPromise;
    return {
      datasources: [
        {
          activeProvider: this.config.activeProvider
        }
      ]
    };
  }
  async uploadSchema() {
    await this.initPromise;
    const res = await request(await this.url("schema"), {
      method: "PUT",
      headers: this.headers,
      body: this.config.inlineSchema
    });
    if (res) {
      this.logEmitter.emit("info", {
        message: `Schema (re)uploaded (hash: ${this.schemaHash})`
      });
    } else {
      this.logEmitter.emit("warn", { message: "Could not upload the schema" });
      throw new Error("Could not upload the schema");
    }
  }
  request(query, headers, attempt = 0) {
    this.logEmitter.emit("query", { query });
    return this.requestInternal({ query, variables: {} }, headers, attempt);
  }
  async requestBatch(queries, headers, isTransaction = false, attempt = 0) {
    this.logEmitter.emit("query", {
      query: `Batch${isTransaction ? " in transaction" : ""} (${queries.length}):
${queries.join("\n")}`
    });
    const body = {
      batch: queries.map((query) => ({ query, variables: {} })),
      transaction: isTransaction
    };
    const { batchResult } = await this.requestInternal(body, headers, attempt);
    return batchResult;
  }
  async requestInternal(body, headers, attempt) {
    await this.initPromise;
    try {
      this.logEmitter.emit("info", {
        message: `Calling ${await this.url("graphql")} (n=${attempt})`
      });
      const res = await request(await this.url("graphql"), {
        method: "POST",
        headers: { ...headers, ...this.headers },
        body: JSON.stringify(body)
      });
      if (res.status === 404) {
        await this.uploadSchema();
        throw new Error("Schema (re)uploaded");
      }
      if (!res.ok) {
        throw new Error("GraphQL request failed");
      }
      return res.json();
    } catch (err) {
      if (attempt >= MAX_RETRIES) {
        this.logEmitter.emit("error", {
          message: `Failed to query: ${err.message}`
        });
        throw err;
      } else {
        const delay = await backOff(attempt);
        this.logEmitter.emit("warn", { message: `Retrying after ${delay}ms` });
        return this.requestInternal(body, headers, attempt + 1);
      }
    }
  }
  transaction() {
    throw new Error("Transactions are currently not supported in Data Proxy");
  }
  extractHostAndApiKey() {
    const mainDatasourceName = Object.keys(this.inlineDatasources)[0];
    const mainDatasource = this.inlineDatasources[mainDatasourceName];
    const mainDatasourceURL = mainDatasource?.url.value;
    const mainDatasourceEnv = mainDatasource?.url.fromEnvVar;
    const loadedEnvURL = this.env[mainDatasourceEnv];
    const dataProxyURL = mainDatasourceURL ?? loadedEnvURL;
    let url;
    try {
      url = new URL(dataProxyURL ?? "");
    } catch {
      throw new Error("Could not parse URL of the datasource");
    }
    const { protocol, host, searchParams } = url;
    if (protocol !== "prisma:") {
      throw new Error("Datasource URL should use prisma:// protocol");
    }
    const apiKey = searchParams.get("api_key");
    if (apiKey === null || apiKey.length < 1) {
      throw new Error("No valid API key found in the datasource URL");
    }
    return [host, apiKey];
  }
}
__name(DataProxyEngine, "DataProxyEngine");
export {
  DataProxyEngine
};
